diff --git a/drivers/staging/fsl_ppfe/pfe_eth.h b/drivers/staging/fsl_ppfe/pfe_eth.h
index 11c1ffb93af8..1014ac43c347 100644
--- a/drivers/staging/fsl_ppfe/pfe_eth.h
+++ b/drivers/staging/fsl_ppfe/pfe_eth.h
@@ -153,7 +153,7 @@ struct  pfe_eth_priv_s {
 };
 
 struct pfe_eth {
-	struct pfe_eth_priv_s *eth_priv[3];
+	struct pfe_eth_priv_s *eth_priv[NUM_GEMAC_SUPPORT];
 };
 
 struct pfe_mdio_priv_s {
