diff --git a/arch/arm64/boot/dts/freescale/fsl-ls1012a.dtsi b/arch/arm64/boot/dts/freescale/fsl-ls1012a.dtsi
index 0ca51cbd1976..59748f654a92 100644
--- a/arch/arm64/boot/dts/freescale/fsl-ls1012a.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-ls1012a.dtsi
@@ -452,6 +452,8 @@
 			snps,dis_rxdet_inp3_quirk;
 			snps,incr-burst-type-adjustment = <1>, <4>, <8>, <16>;
 			snps,host-vbus-glitches;
+			clock-names = "ref", "bus_early", "suspend";
+			clocks = <&clockgen 5 0>, <&clockgen 5 0>, <&clockgen 5 0>;
 		};
 
 		sata: sata@3200000 {
