diff --git a/drivers/net/ethernet/micrel/ksz9897/ksz9897.h b/drivers/net/ethernet/micrel/ksz9897/ksz9897.h
index 8352651..28b21a1 100644
--- a/drivers/net/ethernet/micrel/ksz9897/ksz9897.h
+++ b/drivers/net/ethernet/micrel/ksz9897/ksz9897.h
@@ -1020,6 +1020,10 @@
 
 #define MMD_DEVICE_ID_COMMON		2
 
+#define MMD_LED_MODE			0x00
+#define LED_MODE_SINGLE			(0 << 4)
+#define LED_MODE_TRI_COLOR		(1 << 4)
+
 #define MMD_DEVICE_ID_EEE_ADV		7
 
 #define MMD_EEE_ADV			0x3C
diff --git a/drivers/net/ethernet/micrel/ksz9897/ksz_sw_9897.c b/drivers/net/ethernet/micrel/ksz9897/ksz_sw_9897.c
index ec68241..9900423 100644
--- a/drivers/net/ethernet/micrel/ksz9897/ksz_sw_9897.c
+++ b/drivers/net/ethernet/micrel/ksz9897/ksz_sw_9897.c
@@ -8730,6 +8730,11 @@ static void sw_setup(struct ksz_sw *sw)
 			&val, 1);
 #endif
 
+		val = LED_MODE_SINGLE;
+		port_mmd_write(sw, port, MMD_DEVICE_ID_COMMON, MMD_LED_MODE,
+			&val, 1);
+
+
 		/*
 		 * Switch actually cannot do auto-negotiation with old 10Mbit
 		 * hub.
