diff --git a/arch/arm64/boot/dts/freescale/fsl-ls1012a-nano.dts b/arch/arm64/boot/dts/freescale/fsl-ls1012a-nano.dts
new file mode 100644
index 0000000..e310f57
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/fsl-ls1012a-nano.dts
@@ -0,0 +1,219 @@
+/*
+ * Device Tree file for INCOstartec NanoFW Board.
+ *
+ * Copyright 2016 Freescale Semiconductor, Inc.
+ * Copyright 2018 Telco-Tech GmbH
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPLv2 or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This library is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This library is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+/dts-v1/;
+
+#include "fsl-ls1012a.dtsi"
+#include <dt-bindings/net/ti-dp83867.h>
+
+/ {
+	model = "LS1012A Nano Board";
+	compatible = "fsl,ls1012a-nano", "fsl,ls1012a";
+
+	aliases {
+		serial0 = &duart0;
+		ethernet0 = &pfe_mac0;
+		ethernet1 = &pfe_mac1;
+	};
+};
+
+&duart0 {
+	clocks = <&clockgen 4 3>;
+	status = "okay";
+};
+
+&i2c0 {
+	clocks = <&clockgen 4 3>;
+	clock-frequency = <400000>;
+	status = "okay";
+
+	pca9555: pca9555@20 {
+		compatible = "nxp,pca9555";
+		reg = <0x20>;
+		status = "okay";
+	};
+
+	mcp7941x: mcp7941x@6f {
+		compatible = "microchip,mcp7941x";
+		reg = <0x6f>;
+		status = "okay";
+	};
+
+	ksz9897: ksz9897@5f {
+		compatible = "microchip,ksz9897";
+		reg = <0x5f>;
+		sw_host_port = <7>;
+		status = "okay";
+
+		mdio@0 { /* sw.0 */
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			reg = <0x1>;
+
+			port1: ethernet-phy@0 {
+				reg = <0>;
+			};
+			port2: ethernet-phy@1 {
+				reg = <1>;
+			};
+			port3: ethernet-phy@2 {
+				reg = <2>;
+			};
+			port4: ethernet-phy@3 {
+				reg = <3>;
+			};
+			port5: ethernet-phy@4 {
+				reg = <4>;
+			};
+
+			/* port6 not connected */
+
+			port7: ethernet-phy@6 {	/* u-boot may remove this */
+				reg = <6>;
+			};
+		};
+	};
+};
+
+&qspi {
+	status = "okay";
+
+	qflash0: is25wp128@0 {
+		compatible = "spansion,m25p80";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		spi-max-frequency = <20000000>;
+		m25p,fast-read;
+		reg = <0>;
+		spi-rx-bus-width = <2>;
+		spi-tx-bus-width = <2>;
+	};
+};
+
+&sata {
+	clocks = <&clockgen 4 3>;
+	status = "okay";
+};
+
+&esdhc0 {
+	sd-uhs-sdr25;
+	sd-uhs-sdr12;
+	voltage-ranges = <3300 3300>;
+	cd-gpios = <&gpio0 21 1>;
+	disable-wp;
+	no-1-8-v;
+	status = "okay";
+};
+
+&esdhc1 {
+	voltage-ranges = <1800 1800>;
+	mmc-hs200-1_8v;
+	non-removable;
+	status = "okay";
+};
+
+&pfe {
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	/* NOTE
+	 * u-boot may change the setup for existing hardware
+	 * see ls1012anano.c board setup file (u-boot)
+	 */
+
+	pfe_mac0: ethernet@0 {
+		compatible = "fsl,pfe-gemac-port";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x0>;			/* GEM_ID */
+		fsl,gemac-bus-id = <0x0>;	/* BUS_ID */
+		fsl,gemac-phy-id = <0x2>;	/* PHY_ID */
+		fsl,mdio-mux-val = <0x0>;
+		phy-mode = "sgmii";
+		phy-handle = <&sgmii_phy1>;
+		fsl,pfe-phy-if-flags = <0x0>;
+	};
+
+	pfe_mac1: ethernet@1 {
+		compatible = "fsl,pfe-gemac-port";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x1>;			/* GEM_ID */
+		fsl,gemac-bus-id = <0x1>;	/* BUS_ID */
+		fsl,gemac-phy-id = <0x1>;	/* PHY_ID */
+		fsl,mdio-mux-val = <0x0>;
+		phy-mode = "rgmii-txid";	/* u-boot may change this */
+		phy-handle = <&rgmii_phy2>;	/* u-boot may change this */
+		fsl,pfe-phy-if-flags = <0x0>;
+	};
+
+	mdio@0 {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reg = <0x1>;		/* enabled/disabled */
+
+		sgmii_phy1: ethernet-phy@2 {
+			compatible = "ethernet-phy-id2000.a231";
+			reg = <2>;
+			ti,fifo-depth = <DP83867_PHYCR_FIFO_DEPTH_4_B_NIB>;
+			ti,led-1-source = DP83867_LEDCR_RECEIVE_OR_TRANSMIT_ACTIVITY;
+			ti,pwdn_int;
+		};
+
+		/* NOTE this is the phy from gamc 1 - connected to bus MDIO 0 */
+		rgmii_phy2: ethernet-phy@1 {	/* u-boot may remove this */
+			compatible = "ethernet-phy-id2000.a231";
+			reg = <1>;
+			ti,rx-internal-delay = <DP83867_RGMIIDCTL_2_00_NS>;
+			ti,tx-internal-delay = <DP83867_RGMIIDCTL_2_00_NS>;
+			ti,fifo-depth = <DP83867_PHYCR_FIFO_DEPTH_4_B_NIB>;
+			ti,led-1-source = DP83867_LEDCR_RECEIVE_OR_TRANSMIT_ACTIVITY;
+			ti,pwdn_int;
+		};
+	};
+
+};
+
