diff --git a/lib/efi_loader/efi_disk.c b/lib/efi_loader/efi_disk.c
index c037526ad2..68cdf7e203 100644
--- a/lib/efi_loader/efi_disk.c
+++ b/lib/efi_loader/efi_disk.c
@@ -412,6 +412,9 @@ efi_status_t efi_disk_register(void)
 			if (desc->type == DEV_TYPE_UNKNOWN)
 				continue;
 
+			if (cur_drvr->select_hwpart != NULL)
+				cur_drvr->select_hwpart(desc, 0);
+
 			snprintf(devname, sizeof(devname), "%s%d",
 				 if_typename, i);
 
