diff --git a/drivers/ata/ahci_qoriq.c b/drivers/ata/ahci_qoriq.c
index 0048880889ef..68e03d4fe2c1 100644
--- a/drivers/ata/ahci_qoriq.c
+++ b/drivers/ata/ahci_qoriq.c
@@ -181,6 +181,7 @@ static struct scsi_host_template ahci_qoriq_sht = {
 	AHCI_SHT(DRV_NAME),
 };
 
+#ifdef CONFIG_AHCI_QORIQ
 void fsl_sata_errata_379364(struct ata_link *link)
 {
 	struct ata_port *ap = link->ap;
@@ -311,7 +312,7 @@ void fsl_sata_errata_379364(struct ata_link *link)
 dev_unmap:
 	iounmap(dev_con_base);
 }
-
+#endif
 
 static int ahci_qoriq_phy_init(struct ahci_host_priv *hpriv)
 {
