diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
index a7680fbbeef0..314be074dd5a 100644
--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -1005,6 +1005,9 @@ static const struct flash_info spi_nor_ids[] = {
 
 	/* ISSI */
 	{ "is25cd512", INFO(0x7f9d20, 0, 32 * 1024,   2, SECT_4K) },
+	{ "is25wp32",  INFO(0x9d7016, 0, 64 * 1024,  64, 0) },
+	{ "is25wp64",  INFO(0x9d7017, 0, 64 * 1024, 128, 0) },
+	{ "is25wp128", INFO(0x9d7018, 0, 64 * 1024, 256, 0) },
 
 	/* Macronix */
 	{ "mx25l512e",   INFO(0xc22010, 0, 64 * 1024,   1, SECT_4K) },
