diff --git a/drivers/net/phy/dp83867.c b/drivers/net/phy/dp83867.c
index e4bf9e7d7583..434e5b248021 100644
--- a/drivers/net/phy/dp83867.c
+++ b/drivers/net/phy/dp83867.c
@@ -28,6 +28,7 @@
 #define MII_DP83867_PHYCTRL	0x10
 #define MII_DP83867_MICR	0x12
 #define MII_DP83867_ISR		0x13
+#define MII_DP83867_LEDCR1	0x18
 #define DP83867_CTRL		0x1f
 #define DP83867_CFG3		0x1e
 
@@ -91,6 +92,11 @@ struct dp83867_private {
 	int rx_id_delay;
 	int tx_id_delay;
 	int fifo_depth;
+	uint8_t led_0_source;
+	uint8_t led_1_source;
+	uint8_t led_2_source;
+	uint8_t led_gpio_source;
+	uint32_t led_mask;	
 	int io_impedance;
 	int port_mirroring;
 	bool rxctrl_strap_quirk;
@@ -179,6 +185,29 @@ static int dp83867_of_init(struct phy_device *phydev)
 	dp83867->rxctrl_strap_quirk = of_property_read_bool(of_node,
 					"ti,dp83867-rxctrl-strap-quirk");
 
+	dp83867->led_mask = ~0;
+
+	ret = of_property_read_u8(of_node, "ti,led-0-source",
+				   &dp83867->led_0_source);
+	if (ret)
+		dp83867->led_mask &= ~0x000f;
+
+	ret = of_property_read_u8(of_node, "ti,led-1-source",
+				   &dp83867->led_1_source);
+	if (ret)
+		dp83867->led_mask &= ~0x00f0;
+
+	ret = of_property_read_u8(of_node, "ti,led-2-source",
+				   &dp83867->led_2_source);
+	if (ret)
+		dp83867->led_mask &= ~0x0f00;
+
+	ret = of_property_read_u8(of_node, "ti,led-gpio-source",
+				   &dp83867->led_gpio_source);
+	if (ret)
+		dp83867->led_mask &= ~0xf000;
+
+
 	ret = of_property_read_u32(of_node, "ti,rx-internal-delay",
 				   &dp83867->rx_id_delay);
 	if (ret &&
@@ -312,6 +341,16 @@ static int dp83867_config_init(struct phy_device *phydev)
 		phy_write_mmd(phydev, DP83867_DEVADDR, DP83867_IO_MUX_CFG, val);
 	}
 
+	if (dp83867->led_mask) {
+		ret = dp83867->led_0_source | (dp83867->led_1_source << 4)
+					    | (dp83867->led_2_source << 8)
+					    | (dp83867->led_gpio_source << 12);
+
+		val = phy_read(phydev, MII_DP83867_LEDCR1);
+		val = (val & ~dp83867->led_mask) | (ret & dp83867->led_mask);
+		phy_write(phydev, MII_DP83867_LEDCR1, val);
+	}
+
 	return 0;
 }
 
diff --git a/include/dt-bindings/net/ti-dp83867.h b/include/dt-bindings/net/ti-dp83867.h
index 7b1656427cbe..47a9ab1e21e8 100644
--- a/include/dt-bindings/net/ti-dp83867.h
+++ b/include/dt-bindings/net/ti-dp83867.h
@@ -42,6 +42,21 @@
 #define	DP83867_RGMIIDCTL_3_75_NS	0xe
 #define	DP83867_RGMIIDCTL_4_00_NS	0xf
 
+#define DP83867_LEDCR_LINK_ESTABLISHED				/bits/ 8 <0x00>
+#define DP83867_LEDCR_RECEIVE_OR_TRANSMIT_ACTIVITY		/bits/ 8 <0x01>
+#define DP83867_LEDCR_TRANSMIT_ACTIVITY				/bits/ 8 <0x02>
+#define DP83867_LEDCR_RECEIVE_ACTIVITY				/bits/ 8 <0x03>
+#define DP83867_LEDCR_COLLISION_DETECTED			/bits/ 8 <0x04>
+#define DP83867_LEDCR_1000BT_LINK_ESTABLISHED			/bits/ 8 <0x05>
+#define DP83867_LEDCR_100BTX_LINK_ESTABLISHED			/bits/ 8 <0x06>
+#define DP83867_LEDCR_10BT_LINK_ESTABLISHED			/bits/ 8 <0x07>
+#define DP83867_LEDCR_10_100BT_LINK_ESTABLISHED			/bits/ 8 <0x08>
+#define DP83867_LEDCR_100_1000BT_LINK_ESTABLISHED		/bits/ 8 <0x09>
+#define DP83867_LEDCR_FULL_DUPLEX				/bits/ 8 <0x0a>
+#define DP83867_LEDCR_LINK_ESTABLISHED_BLINK_FOR_ACTIVITY	/bits/ 8 <0x0b>
+#define DP83867_LEDCR_RECEIVE_OR_TRANSMIT_ERROR			/bits/ 8 <0x0d>
+#define DP83867_LEDCR_RECEIVE_ERROR				/bits/ 8 <0x0e>
+
 /* IO_MUX_CFG - Clock output selection */
 #define DP83867_CLK_O_SEL_CHN_A_RCLK		0x0
 #define DP83867_CLK_O_SEL_CHN_B_RCLK		0x1
