diff --git a/drivers/staging/fsl_ppfe/pfe_hif.h b/drivers/staging/fsl_ppfe/pfe_hif.h
index fc4c6c08f094..ac827a927bb9 100644
--- a/drivers/staging/fsl_ppfe/pfe_hif.h
+++ b/drivers/staging/fsl_ppfe/pfe_hif.h
@@ -35,7 +35,7 @@ enum {
  * HIF_TX_DESC_NT value should be always greter than 4,
  * Otherwise HIF_TX_POLL_MARK will become zero.
  */
-#define HIF_RX_DESC_NT		256
+#define HIF_RX_DESC_NT		2048
 #define HIF_TX_DESC_NT		2048
 
 #define HIF_FIRST_BUFFER	BIT(0)
