diff --git a/drivers/staging/fsl_ppfe/pfe_ls1012a_platform.c b/drivers/staging/fsl_ppfe/pfe_ls1012a_platform.c
index bf4b8dd7a676..69d6f7dfc02b 100644
--- a/drivers/staging/fsl_ppfe/pfe_ls1012a_platform.c
+++ b/drivers/staging/fsl_ppfe/pfe_ls1012a_platform.c
@@ -68,7 +68,7 @@ static int pfe_get_gemac_if_properties(struct device_node *gem,
 		} else {
 			phy_id = be32_to_cpup(addr);
 			pdata->ls1012a_eth_pdata[port].phy_id = phy_id;
-			pdata->ls1012a_mdio_pdata[0].phy_mask &= ~(1 << phy_id);
+			pdata->ls1012a_mdio_pdata[port].phy_mask = ~(1 << phy_id);
 		}
 
 		/* If PHY is enabled, read mdio properties */
@@ -200,8 +200,6 @@ static int pfe_platform_probe(struct platform_device *pdev)
 		goto err_prop;
 	}
 
-	pfe_platform_data.ls1012a_mdio_pdata[0].phy_mask = 0xffffffff;
-
 	for (ii = 0; ii < interface_count; ii++) {
 		gem = of_get_next_child(np, gem);
 		if (gem)
